server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java

```diff
@@ -470,6 +474,31 @@ protected void updateIndicesReadOnly(Set<String> indicesToUpdate, ActionListener
             .execute(wrappedListener.map(r -> null));
     }
 
+    private void removeExistingIndexBlocks(ActionListener<Void> listener) {
+        ActionListener<Void> wrappedListener = ActionListener.wrap(listener::onResponse, e -> {
+            logger.debug("removing read-only blocks from indices failed", e);
+            listener.onFailure(e);
+        });
+        final ClusterState state = clusterStateSupplier.get();
+        final Set<String> indicesToRelease = state.routingTable()
+            .indicesRouting()
+            .keySet()
+            .stream()
+            .filter(index -> state.getBlocks().hasIndexBlock(index, IndexMetadata.INDEX_READ_ONLY_ALLOW_DELETE_BLOCK))
+            .collect(Collectors.toUnmodifiableSet());
```

Comment:
@DaveCTurner, please see https://github.com/elastic/elasticsearch/pull/87841/commits/9aa743baf84eac60cc8a17a37d2a33f1987b891f.

Reply:
I think this works :) I do agree that it's a little ugly to poll for a condition rather than reacting to events as they occur but really it's not a huge deal in this case and the alternatives seem to be quite a lot more invasive.

Reply:
Yeah. Maybe something to consider for https://github.com/elastic/elasticsearch/issues/88587. I tried [something different](https://github.com/elastic/elasticsearch/pull/88277/commits/c8cf61504b88824ec5ee188373e8f5cc265fc8ae#diff-6fe0b36a03a9b18dbf7c3276af06485d2ec2e5527d61befc87201a448304e06bR11-R15) before, but that would need some support from `InternalClusterInfoService`.

Reply:
I agree, using the flag makes sense. Evenso, I wonder if it is worth iterating over the blocks directly instead. I.e., something like `state.getBlocks().indices().keySet().stream().filter(...)` instead? I'd expect that list to be much smaller than the full list of indices.

Reply:
Going through the routing table rather than the blocks directly is also done in [one more place](https://github.com/elastic/elasticsearch/blob/c0bc85522d7a5bca8d8bd58726d917e716efb32c/server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdMonitor.java#L380) in this file. What you're suggesting would also apply there I think, if the two are equivalent. Would it be ok to move that to a new PR? I can look into it and do it in both places. Or would you rather I do both in the PR?

Reply:
I'd prefer to at least do this one in this PR. I'd be happy to defer the other one to a separate PR.

Reply:
Done in https://github.com/elastic/elasticsearch/pull/87841/commits/142ec694b9219bff1a100d29eeca40a66802b06e.